tests/test262_runner.rs

```diff
@@ -0,0 +1,589 @@
+struct Cli {
+    #[arg(short, long)]
+    update_expectations: bool,
+    #[arg(long = "dont-treat-crashes-as-failures", action = clap::ArgAction::SetFalse)]
+    treat_crashes_as_failures: bool,
+    filters: Vec<PathBuf>,
+}
+
+fn main() {
+    let mut cli = Cli::parse();
+
+    // We're expecting this binary to always be run in the same machine at
+    // the same time as the repo checkout exists.
+    let runner_base_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    let tests_base = runner_base_path.join("test262/test");
+    let nova_harness_path = runner_base_path.join("nova-harness.js");
+
+    let nova_cli_path = {
+        // current_exe should be target/release/test262(.exe)
+        // The nova_cli binary is target/release/nova_cli(.exe)
+        let mut path = std::env::current_exe().unwrap();
+        assert!(path.pop());
+        path.push("nova_cli");
+        if cfg!(windows) {
+            path.set_extension(".exe");
+        }
+        assert!(path.is_file());
+        path
+    };
+
+    let expectation_path = runner_base_path.join("expectations.json");
+    let expectations = {
+        if cli.update_expectations && !expectation_path.is_file() {
+            Default::default()
+        } else {
+            let file = File::open(&expectation_path).unwrap();
+            if cli.update_expectations && file.metadata().unwrap().len() == 0 {
+                Default::default()
+            } else {
+                let read_result = serde_json::from_reader(&file);
+                if cli.update_expectations
+                    && read_result.is_err()
+                    && !read_result.as_ref().unwrap_err().is_io()
+                {
+                    // Failed to parse JSON, but it's okay since we're updating
+                    // the expectations file anyway.
+                    Default::default()
+                } else {
+                    read_result.unwrap()
+                }
+            }
+        }
+    };
+
+    // Preprocess filters
+    let mut filters_are_valid = true;
+    for filter in cli.filters.iter_mut() {
+        let absolute = tests_base.join(&*filter);
+        assert!(absolute.is_absolute());
+        let Ok(canonical) = absolute.canonicalize() else {
+            filters_are_valid = false;
+            break;
+        };
+        assert!(canonical.is_absolute());
+        let Ok(relative) = canonical.strip_prefix(&tests_base) else {
+            filters_are_valid = false;
+            break;
+        };
+        if *filter != relative {
+            *filter = relative.to_path_buf();
+        }
+    }
+
+    let mut runner = Test262Runner {
+        runner_base_path,
+        tests_base,
+        nova_harness_path,
+        nova_cli_path,
+        treat_crashes_as_failures: cli.treat_crashes_as_failures,
+        expectations,
+        num_tests_run: 0,
+        unexpected_results: Default::default(),
+    };
+    if filters_are_valid {
```

Comment:
**suggestion:** If filters are invalid, we should maybe panic with an error message?

Reply:
It seems like filters can be invalid in the following situations:
- The path doesn't exist (e.g.  `harnes` [sic] rather than `harness`).
- One of the intermediate components of the path isn't a directory (e.g. `harness/assert-false.js/test`).
- The final resolved path is outside the `test262/test` directory (e.g. `../test.js`, or `/etc/fstab`)

I think for the path not existing, it probably shouldn't panic. Not too sure about the other ones.